### PR TITLE
[InstCombine] Simplify compare abs(X) and X.

### LINKED ISSUE #72653: `a <= abs(a)` is always true
Alive2 Proof: https://alive2.llvm.org/ce/z/XD9--h
Original program: https://godbolt.org/z/aT8v6G8e9

### PR DESCRIPTION
**Pull request by Z572**
fix https://github.com/llvm/llvm-project/issues/72653
proof: https://alive2.llvm.org/ce/z/LZzZaj

### CODE REVIEW

**Comment by dtcxzyw** [comment]
See also #73189.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7149]:
```suggestion
        llvm_unreachable("Invalid predicate!");
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7112]: Describe the transform in the header comments.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7119]:
```suggestion
      Value *NullValue = Constant::getNullValue(X->getType());
      Value *MinVal = ConstantInt::get(
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7133]:
```suggestion
        auto *SGT = Builder.CreateICmpSGT(X, AllOnesValue);
```
Canonicalize `sge X, 0 -> sgt X, -1`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7136]: The transform doesn't match the alive2 proof:
```
----------------------------------------
define i1 @src_icmp_uge_abs(i4 %arg) {
#0:
  %abs = abs i4 %arg, 1
  %cmp = icmp uge i4 %abs, %arg
  ret i1 %cmp
}
=>
define i1 @tgt_icmp_uge_abs(i4 %arg) {
#0:
  %cmp = icmp sgt i4 %arg, 15
  ret i1 %cmp
}
Transformation seems to be correct!
```

Generally, `is_int_min_poison=1` provides more information. Thus we shouldn't emit more instructions than the case with `is_int_min_poison=0`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7115]: I think we can drop the one-use constraint as we never create more than one instruction.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7129]: You are matching the operands in swapped order here, but keep the old predicate. This will result in a miscompile (as your last test shows).

**Comment by nikic** [comment]
Please rebase over https://github.com/llvm/llvm-project/commit/9d5b0965c43b4e8b3f21c106fe829391b1382277 and move the fold into foldICmpCommutative().

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6877]:
```suggestion
    if (match(Op0, m_Intrinsic<Intrinsic::abs>(m_Value(X), m_Constant(C))) &&
         match(Op1, m_Specific(X))) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6896]: `icmp ule X, C` will be canonicalized into `icmp ult X, C+1`.

File: llvm/test/Transforms/InstCombine/icmp-abs.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-abs.ll:198]: Please add a negative test `abs(X) < Y`:
```
define i1 @icmp_sge_abs_mismatched_op(i4 %arg, i4 %arg2) {
   %abs = call i4 @llvm.abs.i4(i4 %arg, i1 true)
   %cmp = icmp sge i4 %abs, %arg2
   ret i1 %cmp
 }
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
